ops/pebble.py

```diff
@@ -85,10 +86,46 @@ def _parse_timestamp(s):
     return datetime.datetime.strptime(s, '%Y-%m-%dT%H:%M:%S.%f%z')
 
 
-class ServiceError(Exception):
+def _json_loads(s: Union[str, bytes]) -> Dict:
+    """Like json.loads(), but handle str or bytes.
+
+    This is needed because an HTTP response's read() method returns bytes on
+    Python 3.5, and json.load doesn't handle bytes.
+    """
+    if isinstance(s, bytes):
+        s = s.decode('utf-8')
+    return json.loads(s)
+
+
+class Error(Exception):
+    """Base class of most errors raised by the Pebble client."""
+
+
+class PollTimeout(TimeoutError, Error):
+    """Raised when the wait_change() polling times out."""
+
+
+class SocketError(Error):
+    """Raised when the Pebble client can't connect to the socket."""
+
+
+class APIError(Error):
+    """Raised when an HTTP API error occurs talking to the Pebble server."""
+
+    def __init__(self, body, code, status, message):
+        """This shouldn't be instantiated directly."""
+        super().__init__(message)
+        self.body = body
+        self.code = code
+        self.status = status
+        self.message = message
+
+
+class ServiceError(Error):
     """Raised by wait_change() when a service change is ready but has an error."""
```

Comment:
It's generally worth avoiding documenting the call sites of a given type/method/function in its own documentation, as it's either wrong or at least reduces the value of the particular thing. Instead, ideally the docs would document what the thing does, and we'll use it every time that makes sense. To be clear, though, having ServiceError is nice.

In this particular case, there is also be a second point to consider: do we really want to get an error out of wait_change? As a user of this API I'd probably expect the result of this to be a Change, whether it worked or not. It would indeed make sense for things that assume a service to raise an error, though (say, starting a service). Also worth keeping in mind in this context that a Change doesn't have to be about services.

Reply:
It's a good question -- let me look more closely at what Pebble returns here, but I think you're right that maybe we should just return the `Change`.

Reply:
Hmm, I actually removed this and played around with it, but I don't think that leads to nice Python usage, so I reverted. What it means is that if you call `wait_change()` and an error occurs (like the service not starting) it won't raise an exception, and so the caller will probably merrily proceed in their way ... they'd have to go out of their way to check `change.err` on the return value. I'm worried about folks silently not seeing errors.

Alternatively we could have `wait_change` return the change (with err) but the `start_services` / `stop_services` functions look for this and they would raise. But I kinda think `wait_change` should raise as well, so you get a hard fail when an error occurs.

Reply:
Per note above, that still feels problematic, in two different ways.

If you say wait_change(), you asked the API to wait for a change and want a change back. The fact the change reports a situation in which one of its tasks includes an error is not breaking the expectation of waiting for a change to be returned. A change is composed of tasks, and inside that change some things will have succeeded and some will have failed. The developer asked to just wait for the change, though, and without inspection one cannot even tell what in particular has failed. Whatever it was, though, it wasn't waiting for the change that failed.

With start service, that's different. You asked to start a service, and if one of the tasks that start the service could not take place, it seems reasonable to raise an error to the the logic "No, you asked for a service to start, but I could not, here is an error and now it's your problem again". To make API reasonable, you might even have something like ensure_change() that makes sure the change succeeds instead of just waiting, or add an argument with a similar phrasing to wait_change itself. But even if you do that, that second issue still exists: we cannot extrapolate from just a bare wait_change failure that the error was indeed a service error. Either we need to inspect the change, or we'll need to move the raising of ServiceError to a context that we're sure about what the change is because we created it ourselves.

Reply:
That made me think.. maybe we should actually have ChangeError instead of ServiceError so that we can indeed have something like ensure=True or something along those lines, instead of it being too specific to a particular set of methods. We'll soon start having other async operations that may also fail, and there's probably not much value at this point to differentiate between ServiceError vs. ChangeError. We can always go back and add ServiceError if we need to.

Reply:
Sounds good -- I agree with that logic. I've updated this, specifically so that `wait_change()` returns the `Change` object and never raises an error, but `[autostart|start|stop]_services()` look at the `Change.err` field and raise an error if set -- because you've asked them to perform a specific action, and it didn't go through. (I've also renamed `ServiceError` to the more general `ChangeError` for other async operations later.)